x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/settings/TransportUpdateSecuritySettingsAction.java

```diff
@@ -0,0 +1,171 @@
+import org.elasticsearch.cluster.metadata.MetadataUpdateSettingsService;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.security.action.settings.UpdateSecuritySettingsAction;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import static org.elasticsearch.xpack.security.action.settings.TransportGetSecuritySettingsAction.resolveConcreteIndex;
+import static org.elasticsearch.xpack.security.support.SecuritySystemIndices.SECURITY_MAIN_ALIAS;
+import static org.elasticsearch.xpack.security.support.SecuritySystemIndices.SECURITY_PROFILE_ALIAS;
+import static org.elasticsearch.xpack.security.support.SecuritySystemIndices.SECURITY_TOKENS_ALIAS;
+
+public class TransportUpdateSecuritySettingsAction extends TransportMasterNodeAction<
+    UpdateSecuritySettingsAction.Request,
+    AcknowledgedResponse> {
+
+    private final MetadataUpdateSettingsService updateSettingsService;
+
+    @Inject
+    public TransportUpdateSecuritySettingsAction(
+        TransportService transportService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        ActionFilters actionFilters,
+        MetadataUpdateSettingsService metadataUpdateSettingsService,
+        IndexNameExpressionResolver indexNameExpressionResolver
+    ) {
+        super(
+            UpdateSecuritySettingsAction.NAME,
+            transportService,
+            clusterService,
+            threadPool,
+            actionFilters,
+            UpdateSecuritySettingsAction.Request::new,
+            indexNameExpressionResolver,
+            AcknowledgedResponse::readFrom,
+            ThreadPool.Names.SAME
+        );
+        this.updateSettingsService = metadataUpdateSettingsService;
+    }
+
+    @Override
+    protected void masterOperation(
+        Task task,
+        UpdateSecuritySettingsAction.Request request,
+        ClusterState state,
+        ActionListener<AcknowledgedResponse> listener
+    ) throws Exception {
+
+        List<UpdateSettingsClusterStateUpdateRequest> settingsUpdateRequests = Stream.of(
+            updateSettingsForIndex(
+                SECURITY_MAIN_ALIAS,
+                Settings.builder().loadFromMap(request.mainIndexSettings()).build(),
+                request.timeout(),
+                request.masterNodeTimeout(),
+                state
+            ),
+            updateSettingsForIndex(
+                SECURITY_TOKENS_ALIAS,
+                Settings.builder().loadFromMap(request.tokensIndexSettings()).build(),
+                request.timeout(),
```

Comment:
Wondering about the timeout values here: these will be applied individually to each update request, i.e., the overall composite request may end up running longer than the timeout value, or no?

Reply:
Because we use the `GroupedActionListener` and fire off the requests rapid-fire in the `.forEach()` just below this, the timeouts should run concurrently. How this works is a lot clearer after the name change to this function you suggested above.